srml/consensus/src/lib.rs

```diff
@@ -14,7 +14,89 @@
+//! - `remark` - Make some on-chain remark. The origin of this call must be signed.
+//! - `set_heap_pages` - Set the number of pages in the WebAssembly environment's heap.
+//! - `set_code` - Set the new code.
+//! - `set_storage(Vec<KeyValue>)` - Set some items of storage.
+//! 
+//! ### Public
+//! 
+//! See the [module](https://crates.parity.io/srml_consensus/struct.Module.html) for details on publicly available functions.
+//! 
+//! **Note:** When using the publicly exposed functions, you (the runtime developer) are responsible for implementing any necessary checks (e.g. that the sender is the signer) before calling a function that will affect storage.
+//! 
+//! ## Usage
+//! 
+//! ### Prerequisites
+//! 
+//! To use a functionality from the `consensus` module, call the specific Trait or function that you are invoking from the module with the Rust syntax for doing so, ie
+//! ```rust,ignore
+//! impl<T> for consensus::SomeTrait for Module<T> {
+//! 	/// required functions and types for trait included here
+//! 	/// more comprehensive example included below
+//! }
+//! ```
+//! 
+//! Alternatively, to just set the authorities (presumably for testing purposes with a changing validator set)
+//! ```rust,ignore
+//! consensus::set_authorities(&[<authorities>]) // example included below
+//! ```
+//! 
+//! ### Simple Code Snippet
+//! 
+//! Set authorities
+//! ```rust,ignore
+//! consensus::set_authorities(&[UintAuthorityId(4), UintAuthorityId(5), UintAuthorityId(6)])
+//! ```
+//! 
+//! Log changes in the authorities set
+//! ```rust,ignore
+//! consensus::on_finalise(5); // finalize UintAuthorityId(5)
+//! ```
+//! 
+//! ### Example from SRML
+//! 
+//! In `staking`, the `consensus::OnOfflineReport` is implemented to monitor offline reporting amongst validators:
+//! ```rust,ignore
+//! impl<T: Trait> consensus::OnOfflineReport<Vec<u32>> for Module<T> {
+//! 	fn handle_report(reported_indices: Vec<u32>) {
+//! 		for validator_index in reported_indices {
+//! 			let v = <session::Module<T>>::validators()[validator_index as usize].clone();
+//! 			Self::on_offline_validator(v, 1);
+//! 		}
+//! 	}
+//! }
+//! ```
+//! 
+//! In `grandpa`, we use `consensus` to get the set of `next_authorities` before changing this set according to the consensus algorithm (which does not rotate sessions in the *normal* way):
+//! ```rust,ignore
+//! let next_authorities = <consensus::Module<T>>::authorities()
+//! 			.into_iter()
+//! 			.map(|key| (key, 1)) // evenly-weighted.
+//! 			.collect::<Vec<(<T as Trait>::SessionKey, u64)>>();
+//! ```
+//! 
+//! ## Related Modules
+//! 
+//! - `aura`: this module does not relate directly to `consensus`, but serves to manage offline reporting for the Aura consensus algorithm with its own `handle_report` method
+//! - `grandpa`: this module uses `consensus` to track and manage changes in the validator set in the context of the GRANDPA consensus algorithm
```

Comment:
GRANDPA does its own voter-set managing. But it has a mode where it can track the consensus module, if desired.